### ISSUE TITLE
web-sys: support new interface constructors

### ISSUE THREAD

**Issue by alexlapa** [rustwasm/wasm-bindgen#1952]
### Motivation

Web IDL now allows to define constructors in such way:

```webidl
interface InterfaceWithConstructor {
  [Throws]
  constructor(long a);
};
```
[spec](https://heycam.github.io/webidl/#idl-constructors),[example from gecko-dev](https://github.com/mozilla/gecko-dev/commit/1128b8dee8a369710edb28fbd5da961849efa8df#diff-a51e6b1a90d7281993a986bede267468R82-R85).

rustwasm/weedle is already updated but not released (https://github.com/rustwasm/weedle/pull/33)

Major difference is that new constructors may or may not be marked as `[Throws]`. Atm all constructors return `Result<Self, JsValue>`, [doc](https://github.com/rustwasm/wasm-bindgen/blob/62fee13a46b54896310db0395a634aefe3694845/crates/webidl/src/util.rs#L496-L508). 

### Proposed Solution

Binding generator should support new constructors and all web-sys's WebIDL's should be updated. And that is going to be a big BC-break.

### Alternatives

Keep backward compatibility with constructors defined as attributes thus not changing any WebIDL's. 
That would not break BC and allow adding new WebIDL's. 

But, it seems to me that is not spec compliant that every web-sys's interface constructor can throw, since specs explicitly specify whether constructor can throw, and now it is represented in WebIDL's.

**Comment by alexlapa** [comment]
I think can implement this if maintainers would decide how exactly it should be implemented.

**Comment by alexcrichton** [comment]
I think it would be fine to add support for this roughly at any time. If it's a breaking change we'll have to hold off for awhile until the next breaking version is released, otherwise we may need to implement some tricks to make this backwards compatible.

**Comment by Liamolucko** [comment]
Several types seem to be missing constructor bindings due to this, because their webidl defines them using the new syntax which is being ignored.

**Comment by Liamolucko** [comment]
> Major difference is that new constructors may or may not be marked as `[Throws]`. Atm all constructors return `Result<Self, JsValue>`, [doc](https://github.com/rustwasm/wasm-bindgen/blob/62fee13a46b54896310db0395a634aefe3694845/crates/webidl/src/util.rs#L496-L508).

This bit wasn't resolved by #2842. It doesn't seem like `[Throws]` is actually being used for constructors right now, though (e.g., the URL constructor can throw but [isn't annotated with `[Throws]`](https://url.spec.whatwg.org/#url-class)), so it's probably better to keep them as unconditionally throwing for now.
